### PR TITLE
Post/awards

### PR DESCRIPTION
**Pull request by Stanotech**
https://app.clickup.com/t/8694r2qaj

Adding awards feature added. Backend and frontend. Code review needed.

## Summary by Sourcery

Add a new awards feature to the application, allowing users to give awards to posts with different reward levels and anonymity options. Update models, views, forms, and templates to support this feature, and include tests to ensure its functionality and integrity.

New Features:
- Introduce an awards feature allowing users to give awards to posts, with options for anonymous giving and different reward levels.

Enhancements:
- Add a new PostAward model to manage awards given to posts, including fields for the giver, receiver, post, choice, gold value, and anonymity.
- Implement a PostAwardManager to handle querying of anonymous awards and manage award-related logic.

Documentation:
- Update user profile and post detail templates to display awarded icons and gold awards.

Tests:
- Add comprehensive tests for the awards feature, including tests for anonymous awards, award levels, multiple users giving awards, duplicate prevention, and restrictions on awarding one's own post.

### CODE REVIEW

File: core/models.py [hunk 1]
```diff
@@ -186,6 +187,13 @@ def score(self: "Post") -> int:
     def get_content_type(self: "Post") -> ContentType:
         return ContentType.objects.get_for_model(self)
 
+    def get_post_awards(self: "Post") -> QuerySet:
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/models.py:190]: Corrected

File: core/models.py [hunk 2]
```diff
@@ -268,6 +276,60 @@ def save(self: "PostVote", *args: int, **kwargs: int) -> None:
...
+    anonymous = models.BooleanField(default=False)
+    comment = models.CharField(max_length=100, blank=True)
+
+    class Meta:
+        unique_together: ClassVar[list[str]] = ["post", "user"]
+
+    def __str__(self: "PostVote") -> str:
+        return f"@{self.user}: {self.choice} for post: {self.post}"
+
+    def save(self: "PostVote", *args: int, **kwargs: int) -> None:
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/models.py]: done

File: core/models.py [hunk 3]
```diff
@@ -268,6 +276,60 @@ def save(self: "PostVote", *args: int, **kwargs: int) -> None:
...
+        error_message = "Already given award"
+
+        if self.check_duplicate_award():
+            raise ValueError(error_message)
+
+        if self.choice.startswith("1"):
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/models.py]: Done

File: core/models.py [hunk 4]
```diff
@@ -268,6 +276,60 @@ def save(self: "PostVote", *args: int, **kwargs: int) -> None:
...
+    choice = models.CharField(max_length=20, choices=REWARD_CHOICES)
+    gold = models.IntegerField(default=0)
+    created_at = models.DateTimeField(auto_now_add=True)
+    anonymous = models.BooleanField(default=False)
+    comment = models.CharField(max_length=100, blank=True)
+
+    class Meta:
+        unique_together: ClassVar[list[str]] = ["post", "user"]
+
+    def __str__(self: "PostVote") -> str:
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/models.py]: Done

File: core/models.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/models.py:211]: Done

File: core/forms.py [hunk 5]
```diff
@@ -45,10 +44,31 @@ def __init__(self: "Post", *args: tuple, **kwargs: dict) -> None:
...
+        fields: ClassVar[list[str]] = ["choice", "anonymous", "comment"]
+        widgets: ClassVar[dict[str, forms.Widget]] = {
+            "choice": IconRadioSelect(attrs={"class": "form-check-input"}),
+            "anonymous": forms.CheckboxInput(attrs={"class": "form-check-input"}),
+            "comment": forms.Textarea(attrs={"class": "form-control", "rows": 3, "placeholder": "Optional comment"}),
+        }
+
+        labels: ClassVar[dict[str, str]] = {"choice": ""}
+
+    def __init__(self: "PostAwardForm", *args: list, **kwargs: dict) -> None:
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/forms.py]: Help

File: core/tests/test_awards.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+def test_post_award_multiple_users() -> None:
+    community = Community.objects.create(name="Test community")
+    password = fake.password()
+    user1 = User.objects.create_user(email="testuser1@example.com", nickname="testuser1", password=password)
+    user1.username = "testuser1"
+    user1.community = community
+    user1.save()
+    user2 = User.objects.create_user(email="testuser2@example.com", nickname="testuser2", password=password)
+    user2.username = "testuser2"
+    user2.community = community
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/tests/test_awards.py]: Done

**Comment by Stanotech** [comment]
Adding awards feature added. Backend and frontend. Code review needed.

**Comment by Stanotech** [comment]
sourcery-ai review

**Comment by Stanotech** [comment]
@sourcery-ai review

**Comment by Stanotech** [comment]
@sourcery-ai review

File: core/models.py [hunk 7]
```diff
@@ -268,6 +279,63 @@ def save(self: "PostVote", *args: int, **kwargs: int) -> None:
...
+
+        profile = self.post.author.profile
+        profile.gold_awards += self.gold
+        profile.save(update_fields=["gold_awards"])  # adding gold to profile
+
+        Post.objects.filter(pk=self.post.pk).update(gold=F("gold") + self.gold)  # adding gold to post
+
+        super().save(*args, **kwargs)
+
+    def check_duplicate_award(self: "PostAward") -> bool:
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/models.py:335]: @sourcery-ai  but That's exactly how it is now

**Comment by Stanotech** [comment]
@sourcery-ai review

File: core/models.py [hunk 8]
```diff
@@ -186,6 +194,9 @@ def score(self: "Post") -> int:
     def get_content_type(self: "Post") -> ContentType:
         return ContentType.objects.get_for_model(self)
 
+    def get_post_awards(self: "Post") -> QuerySet:
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/models.py:211]: done

**Comment by Stanotech** [comment]
@sourcery-ai review

File: core/models.py [hunk 9]
```diff
@@ -268,6 +279,57 @@ def save(self: "PostVote", *args: int, **kwargs: int) -> None:
...
+class PostAward(models.Model):
+    REWARD_POINTS: ClassVar[dict[int, int]] = {
+        1: 15,  # Level 1 gives 15 points
+        2: 25,  # Level 2 gives 25 points
+        3: 50,  # Level 3 gives 50 points
+    }
+
+    REWARD_CHOICES: ClassVar[list[tuple[str, str]]] = []
+
+    for level, points in REWARD_POINTS.items():
```
Comment (jacoor) [inline core/models.py]: Please move choice generation into callable as example `get_currencies` in the documentation
https://docs.djangoproject.com/en/5.1/ref/models/fields/#choices

File: core/models.py [hunk 10]
```diff
@@ -268,6 +279,57 @@ def save(self: "PostVote", *args: int, **kwargs: int) -> None:
...
+        unique_together: ClassVar[list[str]] = ["post", "user"]
+
+    def __str__(self: "PostAward") -> str:
+        return f"@{self.user}: {self.choice} for post: {self.post}"
+
+    def save(self: "PostAward", *args: int, **kwargs: int) -> None:
+        if self.check_duplicate_award():
+            raise ValueError("Already given award")
+
+        if self.choice.startswith("1"):
```
Comment (jacoor) [inline core/models.py]: You can try match_case. But this is good.

File: core/forms.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (jacoor) [reply core/forms.py:63]: I am thinking 
```
self.fields["choice"].choices = [choice for choice in self.fields["choice"].choices if choice[0] != ""]
```
can be removed when you move the choices to callable (check comment)

File: core/models.py [hunk 11]
```diff
@@ -268,6 +279,57 @@ def save(self: "PostVote", *args: int, **kwargs: int) -> None:
...
+
+    class Meta:
+        unique_together: ClassVar[list[str]] = ["post", "user"]
+
+    def __str__(self: "PostAward") -> str:
+        return f"@{self.user}: {self.choice} for post: {self.post}"
+
+    def save(self: "PostAward", *args: int, **kwargs: int) -> None:
+        if self.check_duplicate_award():
+            raise ValueError("Already given award")
```
Comment (jacoor) [inline core/models.py]: Won't it already fail on unique together?
Also the check should probably be in the form as I am not sure it will not throw 500 instead of nice form error. Please double check.

File: core/templates/core/icon_radio_select.html [hunk 12]
```diff
@@ -0,0 +1,18 @@
...
+        {% if forloop.parentloop.counter == 1 or forloop.parentloop.counter == 6 or forloop.parentloop.counter == 11 %}
+            {{ option.label }}
+            <br>
+        {% endif %}
+        <label class="form-check-label" for="{{ option.id_for_label }}">
+            <input type="radio" id="{{ option.id_for_label }}" name="{{ widget.name }}" value="{{ option.value }}" class="{{ widget.attrs.class }}">
+            <img src="{% static 'core/icons/' %}{{ option.value }}.png" alt="{{ option.label }}" width="30" height="30">
+        </label>
+
+        {% if forloop.parentloop.counter|divisibleby:5 %}
```
Comment (jacoor) [inline core/templates/core/icon_radio_select.html:13]: Nice!

File: core/tests/test_awards.py [hunk 13]
```diff
@@ -0,0 +1,145 @@
+# Standard library imports
+import pytest
+
+# Django and third-party imports
+from django.contrib.auth import get_user_model
+from faker import Faker
+
+# Local application imports
+from core.models import Community, Post, PostAward
+
```
[the remark this answers is not in the document]
Comment (jacoor) [reply core/tests/test_awards.py:10]: @Stanotech please review sourcery comments and resolve implemented and irrelevant ones.

File: core/tests/test_awards.py [hunk 14]
```diff
@@ -0,0 +1,158 @@
...
+User = get_user_model()
+fake = Faker()
+
+
+@pytest.mark.django_db()
+def test_post_award_level_1() -> None:
```
Comment (jacoor) [inline core/tests/test_awards.py]: You could use already defined fixtures here. Please look at the other tests and conftest.py file. 
Problem is not using fixtures makes those tests hard to maintain, and you have a lot of duplicated code. 
Let me know when done so I can review tests. Thank you.

File: core/views.py [hunk 15]
```diff
@@ -103,6 +103,31 @@ def post(self: "PostVoteView", request: HttpRequest, pk: int, vote_type: str) ->
...
+        context["post"] = post
+        return context
+
+    def form_valid(self: "PostAwardCreateView", form: PostAwardForm) -> HttpResponseRedirect:
+        form.instance.user = self.request.user  # Set current user
+        form.instance.post = get_object_or_404(Post, pk=self.kwargs["pk"])  # Get post with pk from URL
+
+        try:
+            return super().form_valid(form)
+        except DuplicateAwardError as e:
```
Comment (jacoor) [inline core/views.py]: Ah, here is catching the exception. Good one! Disregard my comment on that exception in models.py

File: users/views.py [hunk 16]
```diff
@@ -30,6 +30,13 @@ class UserProfileView(LoginRequiredMixin, DetailView):
     def get_object(self: "UserProfileView") -> User:
         return self.request.user
 
+    def get_context_data(self: "UserProfileView", **kwargs: dict[str, Any]) -> dict[str, Any]:
```
Comment (jacoor) [inline users/views.py]: Do we need that? It can be retrieved in template no? Using relations.

**Review by jacoor — CHANGES_REQUESTED** [review]
Looks good, but tests need to use fixtures. And some small other changes.

File: core/forms.py [hunk 17]
```diff
@@ -45,10 +44,31 @@ def __init__(self: "Post", *args: tuple, **kwargs: dict) -> None:
...
 
 
+class IconRadioSelect(forms.RadioSelect):
+    template_name = "core/icon_radio_select.html"
+
+
+class PostAwardForm(forms.ModelForm):
+    class Meta:
+        model = PostAward
+        fields: ClassVar[list[str]] = ["choice", "anonymous", "comment"]
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/forms.py:54]: done

File: core/tests/test_awards.py [hunk 18]
```diff
@@ -0,0 +1,145 @@
...
+        assert post.post_awards.count() == 1
+        assert user.awards.count() == 1
+        if choice == PostAward.REWARD_CHOICES[0][0]:
+            assert award.gold == 15
+        elif choice == PostAward.REWARD_CHOICES[5][0]:
+            assert award.gold == 25
+        elif choice == PostAward.REWARD_CHOICES[10][0]:
+            assert award.gold == 50
+        # usuń nagrodę, aby móc utworzyć następną
+        award.delete()
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/tests/test_awards.py]: done

File: core/tests/test_awards.py [hunk 19]
```diff
@@ -0,0 +1,145 @@
...
+    post = Post.objects.create(author=user, title="Test post", community=community)
+    award = PostAward.objects.create(post=post, user=user, choice=PostAward.REWARD_CHOICES[0][0])
+
+    assert award.post == post
+    assert award.user == user
+    assert post.post_awards.count() == 1
+    assert user.awards.count() == 1
+    assert award.gold == 15
+    assert post.gold == 15
+    assert user.profile.gold_awards == 1
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/tests/test_awards.py]: done

File: core/models.py [hunk 20]
```diff
@@ -268,6 +279,59 @@ def save(self: "PostVote", *args: int, **kwargs: int) -> None:
...
+    anonymous = models.BooleanField(default=False)
+    comment = models.CharField(max_length=100, blank=True, default="")
+
+    class Meta:
+        unique_together: ClassVar[list[str]] = ["post", "user"]
+
+    def __str__(self: "PostAward") -> str:
+        return f"@{self.user}: {self.choice} for post: {self.post}"
+
+    def save(self: "PostVote", *args: int, **kwargs: int) -> None:
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/models.py]: done

File: core/tests/test_awards.py [hunk 21]
```diff
@@ -0,0 +1,145 @@
...
+    password = fake.password()
+    user1 = User.objects.create_user(email="test@example.com", password=password, nickname="testuser1")
+    user2 = User.objects.create_user(email="test2@example.com", password=password, nickname="testuser2")
+    post = Post.objects.create(author=user1, title="Test post", content="Test content", community=community)
+    PostAward.objects.create(post=post, user=user2, anonymous=True)
+    awards = post.get_post_awards()
+    assert len(awards) == 1
+    assert awards[0].anonymous_nickname == "Anonymous"
+    assert awards[0].user is None
+    assert awards[0].nickname is None
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/tests/test_awards.py]: done

File: core/models.py [hunk 22]
```diff
@@ -186,6 +187,16 @@ def score(self: "Post") -> int:
     def get_content_type(self: "Post") -> ContentType:
         return ContentType.objects.get_for_model(self)
 
+    def get_post_awards(self: "Post") -> QuerySet:
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/models.py:211]: done

File: core/tests/test_awards.py [hunk 23]
```diff
@@ -0,0 +1,141 @@
...
+
+    assert post.post_awards.count() == 3
+    assert user1.awards.count() == 1
+    assert user2.awards.count() == 1
+    assert user3.awards.count() == 1
+    assert award1.gold == 15
+    assert award2.gold == 25
+    assert award3.gold == 50
+    assert post.gold == 15 + 25 + 50
+    assert user1.profile.gold_awards == 15 + 25 + 50
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/tests/test_awards.py]: done

File: core/tests/test_awards.py [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+    user.community = community
+    user.save()
+    post = Post.objects.create(author=user, title="Test post", community=community)
+    award = PostAward.objects.create(post=post, user=user, choice=PostAward.REWARD_CHOICES[0][0])
+
+    assert award.post == post
+    assert award.user == user
+    assert post.post_awards.count() == 1
+    assert user.awards.count() == 1
+    assert award.gold == 15
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/tests/test_awards.py]: done

File: core/tests/test_awards.py [hunk 25]
```diff
@@ -0,0 +1,141 @@
...
+def test_post_award_anonymous() -> None:
+    community = Community.objects.create(name="Test community")
+    password = fake.password()
+    user1 = User.objects.create_user(email="test@example.com", password=password, nickname="testuser1")
+    user2 = User.objects.create_user(email="test2@example.com", password=password, nickname="testuser2")
+    post = Post.objects.create(author=user1, title="Test post", content="Test content", community=community)
+    PostAward.objects.create(post=post, user=user2, anonymous=True)
+    awards = post.get_post_awards()
+    assert len(awards) == 1
+    assert awards[0].anonymous_nickname == "Anonymous"
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/tests/test_awards.py]: done

**Comment by Stanotech** [comment]
@sourcery-ai review

File: core/tests/test_awards.py (hunk 25 above)
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/tests/test_awards.py]: done

[the remark this answers is not in the document]
Comment (Stanotech) [reply core/tests/test_awards.py]: done

File: core/models.py (hunk 9 above)
Comment (Stanotech) [reply core/models.py]: done, but please check if it is correct.
About:
`self.fields["choice"].choices = [choice for choice in self.fields["choice"].choices if choice[0] != ""]`
in forms.py
This code delete first option of IconRadioSelect. I dont know why Django generate one empty option before all awards choices. I checked- there is no empty choice generated. How to prevent django from generating "check nothing" radio select ?

File: core/models.py (hunk 10 above)
Comment (Stanotech) [reply core/models.py]: done

File: core/tests/test_awards.py (hunk 14 above)
Comment (Stanotech) [reply core/tests/test_awards.py]: Done

File: core/models.py (hunk 9 above)
Comment (Stanotech) [reply core/models.py]: Done

File: core/models.py [hunk 26]
```diff
@@ -47,6 +47,26 @@ class ActivePostManagers(PostManagerMixin, ActiveOnlyManager):
     pass
 
 
+class PostAwardManager(models.Manager):
+    def get_post_awards_anonimous(self: "PostAwardManager") -> QuerySet:
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/models.py]: done

File: core/models.py [hunk 27]
```diff
@@ -268,6 +292,60 @@ def save(self: "PostVote", *args: int, **kwargs: int) -> None:
...
 
 
+class PostAward(models.Model):
+    REWARD_POINTS: ClassVar[dict[int, int]] = {
+        1: 15,  # Level 1 gives 15 points
+        2: 25,  # Level 2 gives 25 points
+        3: 50,  # Level 3 gives 50 points
+    }
+
+    REWARD_CHOICES: ClassVar[list[tuple[str, str]]] = []
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/models.py]: done

File: core/forms.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/forms.py]: done

File: users/views.py (hunk 16 above)
Comment (Stanotech) [reply users/views.py]: added "receiver" to model, so now it's not needed

File: core/views.py (hunk 15 above)
Comment (Stanotech) [reply core/views.py]: reorganized. Done.

File: core/tests/test_awards.py [hunk 28]
```diff
@@ -0,0 +1,141 @@
...
+
+# Django and third-party imports
+from django.contrib.auth import get_user_model
+from faker import Faker
+
+# Local application imports
+from core.models import Community, Post, PostAward
+
+User = get_user_model()
+fake = Faker()
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/tests/test_awards.py:16]: done

**Comment by Stanotech** [comment]
@sourcery-ai review

**Comment by Stanotech** [comment]
open once more

**Comment by Stanotech** [comment]
@sourcery-ai review

File: core/models.py [hunk 29]
```diff
@@ -268,6 +292,57 @@ def save(self: "PostVote", *args: int, **kwargs: int) -> None:
...
+
+    objects = PostAwardManager()
+
+    class Meta:
+        unique_together: ClassVar[list[str]] = ["post", "giver"]
+
+    def __str__(self: "PostAward") -> str:
+        return f"@{self.giver}: {self.choice} for post: {self.post}"
+
+    def save(self: "PostAward", *args: int, **kwargs: int) -> None:
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/models.py:324]: does it make sense ?

File: core/models.py [hunk 30]
```diff
@@ -268,6 +292,57 @@ def save(self: "PostVote", *args: int, **kwargs: int) -> None:
...
+    post = models.ForeignKey(Post, on_delete=models.CASCADE, related_name="post_awards")
+    choice = models.CharField(max_length=20, choices=[], blank=True)  # Placeholder for choices
+    gold = models.IntegerField(default=0)
+    created_at = models.DateTimeField(auto_now_add=True)
+    anonymous = models.BooleanField(default=False)
+    comment = models.CharField(max_length=100, blank=True, default="")
+
+    objects = PostAwardManager()
+
+    class Meta:
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/models.py:318]: ruff check made me do this ;C :
: ClassVar[list[str]]
socery-ai  saying that : ClassVar[list[str]] is not necessary (and he is probably right). So How it should be ?

File: core/tests/test_awards.py [hunk 31]
```diff
@@ -0,0 +1,107 @@
...
+
+# Local application imports
+from core.models import Post, PostAward
+
+User = get_user_model()
+fake = Faker()
+
+
+@pytest.mark.django_db()
+def test_post_award_anonymous(users: list[User], post: Post) -> None:
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/tests/test_awards.py:20]: done

File: core/tests/test_awards.py [hunk 32]
```diff
@@ -0,0 +1,107 @@
...
+@pytest.mark.django_db()
+def test_post_award_duplicate_prevention(users: list[User], post: Post, user: User) -> None:
+    post.author = user
+    PostAward.objects.create(
+        receiver=post.author, post=post, giver=users[1], choice=PostAward.get_reward_choices()[0][0]
+    )
+
+    with pytest.raises(
+        IntegrityError, match="UNIQUE constraint failed: core_postaward.post_id, core_postaward.giver_id"
+    ):
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/tests/test_awards.py]: done

File: core/tests/test_awards.py [hunk 33]
```diff
@@ -0,0 +1,107 @@
...
+from django.db import IntegrityError
+from django.test import Client
+from django.urls import reverse
+from faker import Faker
+
+# Local application imports
+from core.models import Post, PostAward
+
+User = get_user_model()
+fake = Faker()
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/tests/test_awards.py]: done

File: core/tests/test_awards.py [hunk 34]
```diff
@@ -0,0 +1,107 @@
...
+
+
+@pytest.mark.django_db()
+def test_cannot_give_award_to_own_post(client: Client, user: User, post: Post) -> None:
+    client.force_login(user)  # log in user
+    award_url = reverse("post-award", kwargs={"pk": post.pk})
+    response = client.post(award_url, {"choice": "1"})  # try to add award to your own post
+
+    assert response.status_code == 302  # check if user is unable to give reward to his own post
+    assert PostAward.objects.count() == 0  # check if user is unable to give reward to his own post
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/tests/test_awards.py]: done

File: core/models.py [hunk 35]
```diff
@@ -268,6 +292,57 @@ def save(self: "PostVote", *args: int, **kwargs: int) -> None:
...
+        reward_points = (
+            (1, 15),
+            (2, 25),
+            (3, 50),
+        )
+        reward_choices = []
+        for level, points in reward_points:
+            for i in range(1, 6):  # 5 icons for each level
+                reward_code = f"{level}{i}_REWARD"
+                reward_points_display = f"{points} points"
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/models.py]: done

**Comment by Stanotech** [comment]
@sourcery-ai review

**Comment by Stanotech** [comment]
updating commits

**Comment by Stanotech** [comment]
@sourcery-ai review

File: core/tests/test_awards.py [hunk 36]
```diff
@@ -0,0 +1,122 @@
...
+    users[0].profile.refresh_from_db()
+    awards = post.get_post_awards()
+    assert len(awards) == 2
+    assert post.author.profile.gold_awards == 30
+    assert awards[0]["giver_anonim"] == "Anonymous"
+    assert awards[1]["giver_anonim"] == "test_user_1"
+
+
+@pytest.mark.django_db()
+@pytest.mark.parametrize(
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/tests/test_awards.py]: There is no way to choice invalid option

File: core/models.py (hunk 29 above)
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/models.py:324]: I like the way it is. And you ?

File: core/models.py [hunk 37]
```diff
@@ -47,6 +47,26 @@ class ActivePostManagers(PostManagerMixin, ActiveOnlyManager):
...
+        return (
+            self.get_queryset()
+            .values(
+                "id",
+                "post_id",
+                "choice",
+                "anonymous",
+            )
+            .annotate(
+                giver_anonim=Case(
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/models.py]: done

File: core/views.py [hunk 38]
```diff
@@ -103,6 +103,40 @@ def post(self: "PostVoteView", request: HttpRequest, pk: int, vote_type: str) ->
...
+    form_class = PostAwardForm
+    template_name = "core/post-award.html"
+
+    def get_context_data(self: "PostAwardCreateView", **kwargs: any) -> dict[str, any]:
+        context = super().get_context_data(**kwargs)
+        post = get_object_or_404(Post, pk=self.kwargs["pk"])
+        context["post"] = post
+        return context
+
+    def dispatch(self: "PostAwardCreateView", request: HttpRequest, *args: any, **kwargs: any) -> HttpResponse:
```
[the remark this answers is not in the document]
Comment (Stanotech) [reply core/views.py:130]: It is clear the way it is with error messages ?

**Comment by Stanotech** [comment]
Review please

File: core/models.py [hunk 39]
```diff
@@ -268,6 +292,57 @@ def save(self: "PostVote", *args: int, **kwargs: int) -> None:
...
+    objects = PostAwardManager()
+
+    class Meta:
+        unique_together: ClassVar[list[str]] = ["post", "giver"]
+
+    def __str__(self: "PostAward") -> str:
+        return f"@{self.giver}: {self.choice} for post: {self.post}"
+
+    def save(self: "PostAward", *args: int, **kwargs: int) -> None:
+        match self.choice[0]:
```
Comment (jacoor) [inline core/models.py]: Seems we would have maintenance issue here. In the code in get_reward_choices there are reward values, and here those are repeated. Please move it to class variable and use that class variable for getting values. 
You could probably just move reward_points from  get_reward_choices to class level. This would make code DRY (don't repeat yourself).

File: core/models.py [hunk 40]
```diff
@@ -268,6 +292,57 @@ def save(self: "PostVote", *args: int, **kwargs: int) -> None:
...
+            (2, 25),
+            (3, 50),
+        )
+        reward_choices = []
+        for level, points in reward_points:
+            reward_points_display = f"{points} points"
+            for i in range(1, 6):  # 5 icons for each level
+                reward_code = f"{level}{i}_REWARD"
+                reward_choices.append((reward_code, reward_points_display))
+        cls.choice.choices = reward_choices  # Assign choices to the class variable
```
Comment (jacoor) [inline core/models.py]: Please remove this comment. It is unnecessary because the code is self explanatory.

File: core/tests/test_awards.py [hunk 41]
```diff
@@ -0,0 +1,122 @@
...
+
+    assert post.post_awards.count() == 3
+    assert users[0].awards_given.count() == 1
+    assert users[1].awards_given.count() == 1
+    assert users[2].awards_given.count() == 1
+    assert award1.gold == 15
+    assert award2.gold == 25
+    assert award3.gold == 50
+    assert post.gold == 15 + 25 + 50
+    assert users[0].profile.gold_awards == 15 + 25 + 50
```
Comment (jacoor) [inline core/tests/test_awards.py:85]: this is good! really nice that you show the math, not simply result.

File: core/tests/test_awards.py [hunk 42]
```diff
@@ -0,0 +1,122 @@
...
+        IntegrityError, match="UNIQUE constraint failed: core_postaward.post_id, core_postaward.giver_id"
+    ):
+        PostAward.objects.create(
+            receiver=post.author, post=post, giver=users[1], choice=PostAward.get_reward_choices()[0][0]
+        )
+
+
+@pytest.mark.django_db()
+def test_cannot_give_award_to_own_post(client: Client, user: User, post: Post) -> None:
+    client.force_login(user)  # log in user
```
Comment (jacoor) [inline core/tests/test_awards.py]: Unnecessary comment

**Review by jacoor — CHANGES_REQUESTED** [review]
Small changes please. Really good job!

**Comment by jacoor** [comment]
Please also resolve conflicts, now we don't know if tests pass or fail.

File: core/tests/test_awards.py (hunk 42 above)
Comment (Stanotech) [reply core/tests/test_awards.py]: done

File: core/tests/test_awards.py (hunk 41 above)
Comment (Stanotech) [reply core/tests/test_awards.py:85]: thx

File: core/models.py (hunk 40 above)
Comment (Stanotech) [reply core/models.py]: done

**Comment by Stanotech** [comment]
.

**Comment by Stanotech** [comment]
.

**Comment by Stanotech** [comment]
.

**Comment by Stanotech** [comment]
please accept requested changes

File: core/views.py [hunk 43]
```diff
@@ -147,7 +147,7 @@ def form_valid(self: "PostAwardCreateView", form: PostAwardForm) -> HttpResponse
         return super().form_valid(form)
 
     def get_success_url(self: "PostAwardCreateView") -> str:
-        return reverse("post-detail", kwargs={"pk": self.kwargs["pk"]})
+        return reverse_lazy("post-detail", kwargs={"pk": self.kwargs["pk"]})
```
Comment (jacoor) [inline core/views.py:150]: yup, prefer reverse_lazy than reverse. I would simply use reverse_lazy everywhere. (only a note)

File: core/models.py [hunk 44]
```diff
@@ -269,6 +293,50 @@ def save(self: "PostVote", *args: int, **kwargs: int) -> None:
...
+    objects = PostAwardManager()
+
+    class Meta:
+        unique_together: ClassVar[list[str]] = ["post", "giver"]
+
+    def __str__(self: "PostAward") -> str:
+        return f"@{self.giver}: {self.choice} for post: {self.post}"
+
+    def save(self: "PostAward", *args: int, **kwargs: int) -> None:
+        # Use the REWARD_POINTS to set the gold value
```
Comment (jacoor) [inline core/models.py]: Unneeded comment. Please remove.

**Review by jacoor — CHANGES_REQUESTED** [review]
Small change needed

File: core/models.py (hunk 39 above)
Comment (Stanotech) [reply core/models.py]: done

File: core/models.py (hunk 44 above)
Comment (Stanotech) [reply core/models.py]: done

**Review by jacoor — APPROVED** [review]
